MekHQ/src/mekhq/campaign/mission/resupplyAndCaches/StarLeagueCache.java

```diff
@@ -0,0 +1,643 @@
+
+        try {
+            for (Unit unit : intactUnits) {
+                List<Part> parts = unit.getParts();
+                for (Part part : parts) {
+                    if (part instanceof MekLocation) {
+                        if (((MekLocation) part).getLoc() == LOC_CT) {
+                            continue;
+                        }
+                    }
+
+                    if (part instanceof TankLocation) {
+                        continue;
+                    }
+
+                    // Is the part too damaged to be salvaged?
+                    if (Compute.randomInt(500) < ruinedChance) {
+                        continue;
+                    }
+
+                    Pair<Unit, Part> pair = new Pair<>(unit, part);
+                    int weight = getDropWeight(pair.getValue());
+                    partsPool.merge(part, weight, Integer::sum);
+                }
+            }
+        } catch (Exception exception) {
+            logger.error("Aborted parts collection.", exception);
+        }
+    }
+
+    public void determineOriginFaction() {
+        final int sphereOfInfluence = 650;
+        final PlanetarySystem contractSystem = contract.getSystem();
+        final double distanceToTerra = contractSystem.getDistanceTo(campaign.getSystemById("Terra"));
+
+        // This is a fallback to better ensure something drops, even if it isn't a SLDF Depot
+        // This value was reached by 'eye-balling' the map of the Inner Sphere
+        if (distanceToTerra > sphereOfInfluence) {
+            List<String> factions = contractSystem.getFactions(FALL_OF_STAR_LEAGUE);
+
+            if (factions.isEmpty()) {
+                didGenerationFail = true;
+            } else {
+                Collections.shuffle(factions);
+                originFaction = Factions.getInstance().getFaction(factions.get(0));
+            }
+        } else {
+            originFaction = Factions.getInstance().getFaction("SL");
+        }
+    }
+
+    private List<Unit> getCacheContents() {
+        Map<Integer, List<Integer>> unitsPresent = buildUnitWeightMap();
+        List<MekSummary> unitSummaries = getUnitSummaries(unitsPresent);
+
+        List<Unit> units = new ArrayList<>();
+        for (MekSummary summary : unitSummaries) {
+            Entity entity = getEntity(summary);
+
+            if (entity == null) {
+                continue;
+            }
+
+            Unit unit = getUnit(entity);
+
+            if (unit != null) {
+                units.add(unit);
+            }
+        }
+
+        return units;
+    }
+
+    @Nullable
+    private Entity getEntity(MekSummary unitData) {
+        try {
+            return new MekFileParser(unitData.getSourceFile(), unitData.getEntryName()).getEntity();
+        } catch (Exception ex) {
+            logger.error("Unable to load entity: {}: {}",
+                unitData.getSourceFile(),
+                unitData.getEntryName(), ex);
+            return null;
+        }
+    }
```

Comment:
We need to stop reusing the exact same code everywhere and just create an "EntityUtils" 🤣

Reply:
I concur, because this is obnoxious.